classes/timeslots/timeslots_table.php

```diff
@@ -0,0 +1,113 @@
+ */
+class timeslots_table extends \table_sql implements \renderable {
+
+    /**
+     * Constructs the table and defines how the data from the SQL query is displayed
+     * @param string $uniqueid ID that uniquely identifies this element on the HTML page
+     * @param \moodle_url $callbackurl URL used for callback for action buttons in the table
+     * @param int $perpage number of entries per page for the table
+     */
+    public function __construct(string $uniqueid, \moodle_url $callbackurl, int $perpage = 50) {
+        parent::__construct($uniqueid);
+
+        $this->define_columns([
+            'id',
+            'start_time',
+            'duration',
+            'observer_id',
+            'action',
+        ]);
+
+        $this->define_headers([
+            get_string('id', 'observation'),
+            get_string('starttime', 'observation'),
+            get_string('duration', 'observation'),
+            get_string('observer_id', 'observation'),
+            get_string('actions', 'observation'),
+        ]);
+
+        $this->pagesize = $perpage;
+
+        $systemcontext = \context_system::instance();
+        $this->context = $systemcontext;
+        $this->collapsible(false);
+        $this->sortable(false, 'list_order', SORT_ASC);
+        $this->pageable(true);
+        $this->is_downloadable(false);
+        $this->define_baseurl($callbackurl);
+    }
+
+    /**
+     * Data converter for the lang_string column
+     * @param mixed $row current row
+     */
+    public function col_lang_string($row) {
+        return(get_string($row->lang_string, 'observation'));
+    }
+
+    /**
+     * Generates an action button for the table
+     * @param string $url base URL for the button to link to
+     * @param int $obsid observation ID to add to the URL
+     * @param int $slotid observation slot ID to add to the url
+     * @param string $action action to add to the URL
+     * @param string $text button text
+     */
+    private function action_button(string $url, int $obsid, int $slotid, string $action, string $text) {
+        return \html_writer::link(
+            new \moodle_url($url, ['id' => $obsid, 'action' => $action, 'slotid' => $slotid]),
+            $text,
+            ['class' => 'btn btn-secondary']
+        );
+    }
+
+    /**
+     * Data converter for the action column
+     * @param mixed $row current row
+     */
+    public function col_action($row) {
+        // Add action buttons.
+        $htmlout = $this->action_button($this->baseurl, $row->obs_id, $row->id, 'edit', get_string('edit', 'observation'));
+        $htmlout .= $this->action_button($this->baseurl, $row->obs_id, $row->id, 'delete', get_string('delete', 'observation'));
+        $htmlout .= $this->action_button($this->baseurl, $row->obs_id, $row->id, 'moveup', get_string('moveup', 'observation'));
+        $htmlout .= $this->action_button($this->baseurl, $row->obs_id, $row->id, 'movedown', get_string('movedown', 'observation'));
```

Comment:
Do timeslots need a move up and down action ? Probably only need delete and edit actions.

Reply:
removed